objects/isn_unifiedgrowingtray.lua

```diff
@@ -0,0 +1,524 @@
+		while (storage.fluid < storage.fluidUse) do
+			if isn_doWaterIntake() == false then return false end
+		end
+	end
+	--Consume some fluid
+	storage.fluid = storage.fluid - storage.fluidUse
+	return true
+end
+
+--Fetch the seed from the storage slot, also does validity check.
+function isn_readContainerSeed()
+	local contents = world.containerItems(entity.id())
+	local seed = contents[storage.seedslot]
+	if seed == nil then return nil end
+	--Reused the validity check here as it's useful to know if the seed is valid in almost every case.
+	local seedConfig = root.itemConfig(seed).config
+	if seedConfig.objectType ~= "farmable" then return nil end
+	if seedConfig.objectName == "sapling" and cHandlesSaplings() ~= true then return nil end
+	return seed
+end
+
+--Reads the currentseed's data.  Return false if there was a problem with the
+--seed/data.
+function isn_readSeedData()
+	if storage.currentseed == nil then return false end
+	local seedConfig = root.itemConfig(storage.currentseed).config
+	storage.stages = 0
+	storage.resetStage = 0
+	storage.stage = {}
+	storage.harvestPool = nil
+	--Tree data
+	storage.hasTree = false
+	storage.stemDrops = nil
+	storage.foliageDrops = nil
+	--pairs() has an undefined order...
+	local index = 1
+	while seedConfig.stages[index] ~= nil do
+		local curStage = seedConfig.stages[index]
+		if type(curStage.duration) == "table" then
+			storage.stages = index
+			local stage = {}
+			stage.min = curStage.duration[1] or 10
+			stage.max = curStage.duration[2] or stage.min
+			storage.stage[index] = stage
+		else
+			if curStage.harvestPool ~= nil then storage.harvestPool = curStage.harvestPool end
+			if curStage.resetToStage ~= nil then storage.resetStage = curStage.resetToStage + 1 end
+			if curStage.tree ~= nil then 
+				if curStage.tree == true then
+					if isn_readSaplingData() == false then return false end
+					storage.harvestPool = ""
+				end
+			end
+		end
+		index = index + 1
+	end
+	if index == 1 then return false end -- seed data was invalid.
+	return true
+end
+
+--Reads sapling data and does minimal verification that we indeed have a valid
+--sapling.
+function isn_readSaplingData()
+	if type(storage.currentseed.parameters) ~= "table" then return false end
+	local stemType = storage.currentseed.parameters.stemName
+	local foliageType = storage.currentseed.parameters.foliageName
+	--Not sure if there is any saplings which only have one or the other of stem/foliage but
+	-- just in case the aspects are separated.
+	if stemType then
+		local stemPath = root.treeStemDirectory(stemType)
+		local stemJson = root.assetJson(stemPath .. stemType .. ".modularstem")
+		storage.stemDrops = stemJson.dropConfig.drops
+		storage.hasTree = true
+	end
+	if foliageType then
+		local foliagePath = root.treeFoliageDirectory(foliageType)
+		local foliageJson = root.assetJson(foliagePath .. foliageType .. ".modularfoliage")
+		storage.foliageDrops = foliageJson.dropConfig.drops
+		storage.hasTree = true
```

Comment:
I'm pretty sure a valid sapling HAS to have a valid stemName.
I'm not 100% sure but back then when the stemName was invalid it resulted either in Generic saplings or PGIs (Perfectly Generic Items, e.g. a failure in essence)

Reply:
hubnester & me just talked about it, we both think both should be valid at the same time

Reply:
Yes the code reads and uses the actual treasurePool/drop tables so FU doesn't have to know about a seed/plant and it's output for trays to work with them.  That was the biggest thing that got me started on this.

It also aims to handle perennials properly, that is most perennials don't have many/any seeds in their drop pool and we are no longer explicitly dropping extra seeds so then (imo) it's better not to just keep eating seeds.  If the goal were to eat seeds it is easy to exclude seeds from the drop outputs.

I should note that a lot of the changes besides drop table/pool (ie timing of water use) fell out of having more data to work with and isn't really something I'm strongly for/against.  More did it because it was in front of me and I was aiming to model how plants grow in the game in dirt.

As for saplings, I believe you are quite correct regarding stems though it gets murky on foliage.  There are some trees which do not drop saplings and I don't know if the reason for that is they don't have foliage or their foliage drop table is empty.  Requiring both is sensible and can be revisited if a counter example (no foliage) is found.